### ISSUE TITLE
different output shape for train/test

### ISSUE THREAD

**Issue by chentingpc** [fchollet/keras#2481]
I am implementing a model that requires last layer's output to be of different shapes during train and test.

I use K.in_training_phase(train_output, test_output) where the shape of test_output is different from train_output. It seems Keras cannot support it for now? If not, is there some quick hack to fix this?

**Comment by fluency03** [comment]
what do you mean by different shapes during train and test exactly?  And why you need this?

**Comment by chentingpc** [comment]
Well, different shapes during train and test means, the shape of last layer's output variable, in training phase, it is A, say (None, 1) or even (1, ), but in test phase, it is B, say (None, 2).

It just come up as quick and dirty implementation for negative sampling based training, like word2vec, as in which the target are sampled words. It is not possible to feed target words as output in Keras, because output can only be accessed in objective functions, but I need the target words to help generate prediction, i.e. scores on target words. So I setup a last layer which computes the loss function during training, but in test, it produces scores for given target words.

**Comment by kylemcdonald** [comment]
I don't think it makes sense to ever have more outputs during test/prediction than during training, though sometimes it makes sense the other way around. If you have more outputs for prediction than during training, it will necessarily be useless output because you haven't trained it to be useful.

If the model you have in mind is similar to word2vec, maybe check out some implementations of word2vec in Keras to see how they solve the problem? Here's one https://github.com/niitsuma/word2vec-keras-in-gensim/blob/master/word2veckeras/word2veckeras.py#L116-L124

**Comment by chentingpc** [comment]
I admit there might be other ways around, depending on the specific cases. But just not sure why the training/test output variable shape has to be fixed the same (in call function, not in get_output_shape_for function)? I haven't been able to locate the code that enforces this.

PS, the word2vec implementation is not exact the original word2vec, as original one use log loss but this one use mse. Though change it to log loss is possible, it can be different loss in other cases involving negative sampling (e.g. triple based hinge loss as in WSABIE), where it may be handy to **define loss in some user defined layer before loss function, and no output is required**. And currently this cannot be done in a clean way.
